search_agg_date_histogram.go

```diff
@@ -0,0 +1,236 @@
+package opensearchtools
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// DateHistogramAggregation buckets documents based on a date interval.
+// An empty DateHistogramAggregation will have several issues with execution:
+//   - the target Field must be non-null and non-empty
+//   - the Interval must be non-null and non-empty
+//
+// For more details see https://opensearch.org/docs/latest/opensearch/bucket-agg/#histogram-date_histogram
+type DateHistogramAggregation struct {
+	// Field to be bucketed
+	Field string
+
+	// MinDocCount is the lower count threshold for a bucket to be included in the results.
+	// Negative counts will be omitted
+	MinDocCount int64
+
+	// Interval string using OpenSearch [date math].
+	// [date math]: https://opensearch.org/docs/latest/opensearch/supported-field-types/date/#date-math
+	Interval string
+
+	// TimeZone, times are stored internally in UTC and by default date histograms are bucketed in UTC.
+	// Set the TimeZone to overwrite this default
+	TimeZone string
+
+	// Order list of [Order]s to sort the aggregation buckets. Default order is _count: desc
+	Order []Order
+
+	// Aggregations sub aggregations for each bucket. Mapped by string label to sub aggregation
+	Aggregations map[string]Aggregation
+}
+
+// NewDateHistogramAggregation instantiates a DateHistogramAggregation targeting
+// the provided field with the provided interval. Sets the MinDocCount to -1 to be
+// omitted in favor of the OpenSearch default.
+func NewDateHistogramAggregation(field, interval string) *DateHistogramAggregation {
+	return &DateHistogramAggregation{
+		Field:        field,
+		MinDocCount:  -1,
+		Interval:     interval,
+		Aggregations: make(map[string]Aggregation),
+	}
+}
+
+// WithMinDocCount the lower count threshold for a bucket to be included in the results
+func (d *DateHistogramAggregation) WithMinDocCount(minCount int64) *DateHistogramAggregation {
+	d.MinDocCount = minCount
+	return d
+}
+
+// WithTimeZone overwriting the default UTC timezone
+func (d *DateHistogramAggregation) WithTimeZone(tz string) *DateHistogramAggregation {
+	d.TimeZone = tz
+	return d
+}
+
+// AddOrder of the returned buckets
+func (d *DateHistogramAggregation) AddOrder(orders ...Order) *DateHistogramAggregation {
+	d.Order = append(d.Order, orders...)
+	return d
+}
+
+// AddSubAggregation to the TermsAggregation with the provided name
+func (d *DateHistogramAggregation) AddSubAggregation(name string, agg Aggregation) BucketAggregation {
+	if d.Aggregations == nil {
+		d.Aggregations = map[string]Aggregation{name: agg}
+	} else {
+		d.Aggregations[name] = agg
+	}
+
+	return d
+}
+
+// ConvertSubAggregations uses the provided converter to convert all the sub aggregations in this TermsAggregation
+func (d *DateHistogramAggregation) ConvertSubAggregations(converter AggregateVersionConverter) (map[string]Aggregation, error) {
```

Comment:
Would it make sense to make this just an exported package function instead of a method on each type?  There's nothing here that's specific to this particular request.

Reply:
I could, but to do it right I think we'd need another method on the `BucketAggregation`, something like `SubAggregations() map[string]Aggregation`, so that it's unified. Otherwise we'd have to have a type switch statement to read the sub aggregations from each different type. I'm not opposed to either, but I didn't necessarily love it so I didn't do it.

Thoughts?

Reply:
Maybe.  I was mostly thinking of eliminating the code duplication for this conversion across all the various operations.  They all have this same `make()` and `for` loop that calls `converter` on each element.

Reply:
What's more ideal for the consumer?
If it's better for the consumer to do a method call, you could have each ConvertSubAggregations() call an un-exported function which takes a BucketAggregation with that `SubAggregations()` method... then maybe you have the ideal API plus a bit less duplication.